server/s3_put_multiobject_copy_action.cc

```diff
@@ -0,0 +1,962 @@
+
+const char DirectiveValueReplace[] = "REPLACE";
+const char DirectiveValueCOPY[] = "COPY";
+
+S3PutMultipartCopyAction::S3PutMultipartCopyAction(
+    std::shared_ptr<S3RequestObject> req, std::shared_ptr<MotrAPI> motr_api,
+    std::shared_ptr<S3ObjectMultipartMetadataFactory> object_mp_meta_factory,
+    std::shared_ptr<S3PartMetadataFactory> part_meta_factory,
+    std::shared_ptr<S3BucketMetadataFactory> bucket_meta_factory,
+    std::shared_ptr<S3ObjectMetadataFactory> object_meta_factory,
+    std::shared_ptr<S3MotrWriterFactory> motr_s3_writer_factory,
+    std::shared_ptr<S3MotrReaderFactory> motr_s3_reader_factory,
+    std::shared_ptr<S3MotrKVSWriterFactory> kv_writer_factory,
+    std::shared_ptr<S3AuthClientFactory> auth_factory)
+    : S3PutObjectActionBase(std::move(req), std::move(bucket_meta_factory),
+                            std::move(object_meta_factory), std::move(motr_api),
+                            std::move(motr_s3_writer_factory),
+                            std::move(kv_writer_factory)){
+  part_number = get_part_number();
+  upload_id = request->get_query_string_value("uploadId");
+  s3_log(S3_LOG_DEBUG, request_id, "%s Ctor\n", __func__);
+  s3_log(S3_LOG_INFO, stripped_request_id,
+         "S3 API: UploadPartCopy. Destination: [%s], Source: [%s]\
+         Part[%d] for UploadId[%s]\n",
+         request->get_object_uri().c_str(), request->get_headers_copysource().c_str(),
+         part_number, upload_id.c_str());
+
+  action_uses_cleanup = true;
+  s3_copy_part_action_state = S3CopyPartActionState::empty;
+  old_object_oid = {0ULL, 0ULL};
+  old_layout_id = -1;
+  layout_id = -1;  // Will be set during create object
+  new_object_oid = {0ULL, 0ULL};
+
+  if (S3Option::get_instance()->is_auth_disabled()) {
+    auth_completed = true;
+  }
+
+  if (object_mp_meta_factory) {
+    object_mp_metadata_factory = object_mp_meta_factory;
+  } else {
+    object_mp_metadata_factory =
+        std::make_shared<S3ObjectMultipartMetadataFactory>();
+  }
+
+  if (part_meta_factory) {
+    part_metadata_factory = part_meta_factory;
+  } else {
+    part_metadata_factory = std::make_shared<S3PartMetadataFactory>();
+  }
+
+  if (motr_s3_writer_factory) {
+    motr_writer_factory = motr_s3_writer_factory;
+  } else {
+    motr_writer_factory = std::make_shared<S3MotrWriterFactory>();
+  }
+
+  if (motr_api) {
+    s3_motr_api = std::move(motr_api);
+  } else {
+    s3_motr_api = std::make_shared<ConcreteMotrAPI>();
+  }
+
+  if (kv_writer_factory) {
+    motr_kv_writer_factory = std::move(kv_writer_factory);
+  } else {
+    motr_kv_writer_factory = std::make_shared<S3MotrKVSWriterFactory>();
+  }
+  
+  if (motr_s3_reader_factory) {
+    motr_reader_factory = std::move(motr_s3_reader_factory);
+  } else {
+    motr_reader_factory = std::make_shared<S3MotrReaderFactory>();
+  }
+
+  S3UriToMotrOID(s3_motr_api, request->get_object_uri().c_str(), request_id,
+                 &new_object_oid);
```

Comment:
we are not considering the return code from this function `S3UriToMotrOID`, so are we validating the generated_object_oid somewhere else??

Reply:
If error code is anything other than 0, it should fail and that is handled in function `S3UriToMotrOID` itself. Also few checks like ensuring generated OID does not fall in `motr` and `S3` reserved range are also there. We can always add more validations, please share if you have some ideas.